### PR TITLE
Fixes #381: Migrates to REST API

### LINKED ISSUE #381: GCI Task: Migrating to REST api
## Objective
Presently the entire Badgeyay system is coupled. The next step is to decouple the system. For this we require separate backend and frontend and both should be independent of each other. The objective of this task is to create a REST API for Badgeyay such that all the functionalities of the project can be accessed over the API.

## Requirements
- A github account
- Knowledge of Git
- Knowledge of python
- Knowledge of flask web frame work.
- Knowledge of REST API, how it works.
- Clone the badgeyay repository.
- Create  a suitable issue for this task here.
- Plan your approach, discuss and take help from your mentors.
- Document your approach, the endpoints you will create and how the endpoints will work.
- After sufficient discussion with your mentors start implementing your plan.
- Ask your mentors to create a separate branch, send a PR over there.
- You may create a single PR or multiple PRs.
- Document all the endpoints you have created.

## Outcome
- A working and almost complete REST API for Badgeyay.
- Tweet a link and screenshot of your code to @fossasia @mariobehling @hpdang @gci and/or share it on Facebook or any other social media

### PR DESCRIPTION
**Pull request by abishekvashok**
Fixes #381 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:
Converts the existing Flask app which was coupled into a back-end which
acts as a REpresentational State Transfer based API. Provides all
functionalities of the app to be provided over the new API.

Adds and modifies related documentation.

The tests have been commented out from running in the CI platforms like
.travis.yml as they are specific to the coupled system and not with the
back-end alone.

**Note:** I have'nt added tests to test the back-end. I just commented out the tests from running via Travis. Also note that codeacy if failing because this branch `rest-api` was not enabled in the dashboard.

Preview:
screenshot_20171226_173448

Link to preview changes: https://radiant-thicket-70883.herokuapp.com/

### CODE REVIEW

**Comment by vaibhavsingh97** [comment]
@abishekvashok Can you please provide demo link

**Comment by abishekvashok** [comment]
@djmgit done.

**Comment by abishekvashok** [comment]
@vaibhavsingh97 @djmgit please try this link as a live preview: https://radiant-thicket-70883.herokuapp.com/

**Comment by abishekvashok** [comment]
Ping @fossasia/bagdeyay-admins

File: app/main.py [hunk 1]
```diff
@@ -67,25 +55,44 @@ def empty_directory():
...
+                'type': response_type,
+                'message': message
+            }
+        ]
+    else:
+        response = [
+            {
+                'type': response_type,
+                'message': message,
+                'download_link': download_link
```
Comment (S2606) [inline app/main.py]: Why is download link being provided instead of the pdf file as response??

**Comment by S2606** [comment]
And also we can create api blueprint for documentation purpose

File: app/main.py (hunk 1 above)
Comment (abishekvashok) [reply app/main.py]: The link is provided just as the rest api specifications and common practice. It is what the the web app delivers as well. We can't explicitly send the file as part of an api :P

Comment (abishekvashok) [reply app/main.py]: The download link is the link to download the pdf file.

**Comment by abishekvashok** [comment]
@S2606 i have will put up a google doc once this is merged. I have updated the readme and guide.html as well.

File: app/main.py (hunk 1 above)
Comment (S2606) [reply app/main.py]: Ok @abishekvashok

Comment (abishekvashok) [reply app/main.py]: Sure @S2606

**Comment by abishekvashok** [comment]
@S2606 @djmgit  I await your review :P

File: app/main.py [hunk 2]
```diff
@@ -156,12 +158,19 @@ def upload():
             traceback.print_exc()
 
         if True:
-            flash(filename.replace('.', '-'), 'success-pdf')
-            os.rename(os.path.join(BADGES_FOLDER + "/" + filename + ".badges.pdf"),
-                      os.path.join(BADGES_FOLDER + "/" + filename.replace('.', '-') + "-badges.pdf"))
+            uri = filename.replace('.', '-') + "-badges.pdf"
+            return output('success', 'pdf generation completed successfully', uri)
```
Comment (ParthS007) [inline app/main.py]: @abishekvashok  I think there's a little typo here `url` should be there in place of `uri` in both lines or have you added it but as you are providing link to download , url should be better .
Else looks fine to me, well done 💯

Comment (abishekvashok) [reply app/main.py]: Rest API specifies to use uri

Comment (abishekvashok) [reply app/main.py]: However I could change that to url @ParthS007

Comment (ParthS007) [reply app/main.py]: Ok , no problem , Let's wait for others to review 👍

**Comment by djmgit** [comment]
@abishekvashok ``` curl -X POST https://radiant-thicket-70883.herokuapp.com/api/v1.0/generate_badges
``` is returning bad request where as I should get error message in form of data because I have not provided any param

**Comment by abishekvashok** [comment]
@djmgit can you recheck now, it appeared because the post request was empty. Added case to handle that now.

File: app/main.py (hunk 2 above)
Comment (abishekvashok) [reply app/main.py]: @ParthS007 I changed it now. :)

**Comment by abishekvashok** [comment]
screenshot_20171228_194758

File: app/main.py [hunk 3]
```diff
@@ -67,25 +55,46 @@ def empty_directory():
...
+
+    if 'file' in request.files:
+        file = request.files['file']
 
-    # If default background is selected
+    # img-default is specified
     if img != '':
         if (img == 'user_defined.png'):
             bg_color = request.form['bg_color']
             do_svg2png(img, 1, bg_color)
```
Comment (djmgit) [inline app/main.py]: @abishekvashok what happens if user does not provide bg_color. Exception will be raised. Please use request.form.get('bg_color') and check for empty string.

**Comment by djmgit** [comment]
@abishekvashok I tried the following request
curl --data "csv=a,b,c,d&bg_color=#ad5656&img-default=user_defined.png" https://radiant-thicket-70883.herokuapp.com/api/v1.0/generate_badges
 If I am not wrong this should generate a single badge without any error. But its returning bad request.

**Comment by djmgit** [comment]
@abishekvashok Please look into this. Actually I just want to make sure that the basic API is working fine before merging this.

**Comment by abishekvashok** [comment]
@djmgit on it.

**Comment by abishekvashok** [comment]
@djmgit  done. Sorry for that. Is this ok now?
screenshot_20171228_213712

**Comment by abishekvashok** [comment]
@djmgit sry about that, fixed everything now!
screenshot_20171228_214520

**Comment by abishekvashok** [comment]
I generated a badge now,
screenshot_20171228_214912

**Comment by vaibhavsingh97** [comment]
Good Work, REST API is working fine
Reference:
```json
{
    "response": [
        {
            "download_link": "/static/badges/badge_4-png-csv-badges.pdf",
            "message": "pdf generation completed successfully",
            "type": "success"
        }
    ]
}
```

**Pull request merged**